File: 73/D.cpp
//encoding:utf-8
//#include <iostream> //c++
//#include<stack> //スタックに関して　最新のデータから捨てる
//#include<queue> //キューに関して　最古のデータから捨てる
//#include<math.h>
//#include<vector>
//#include<algorithm> //sort とかにいる
#include<bits/stdc++.h>//vector,algorithm,que
#include<time.h> //時間にかんして

#define ll long long
#define MOD 1000000007
#define INF (1<<29)
using namespace std;
int main(){
  int N,M,R;
  cin>>N>>M>>R;
  vector<int> r(R);

  for(int i=0;i<R;i++){
    cin>>r[i];
  }
  sort(r.begin(),r.end());
  int ways[M][3];
  for(int i=0;i<M;i++){
    cin>>ways[i][0]>>ways[i][1]>>ways[i][2];
  }
  // warshall_floyd
  int cost[N][N];
  for(int i=0;i<N;i++){
    for(int j=0;j<N;j++){
      cost[i][j]=INF;
    }
  }

  for(int i=0;i<M;i++){
    cost[ways[i][0]-1][ways[i][1]-1] = ways[i][2];
    cost[ways[i][1]-1][ways[i][0]-1] = ways[i][2];
  }

  for(int i=0;i<N;i++){
    for(int j=0;j<N;j++){
      for(int k=0;k<N;k++){
        cost[j][k] = min(cost[j][k],cost[j][i]+cost[i][k]);
      }
    }
  }

  ll ans = INF;
  int cnt = 0;
  do{
    ll tmp = 0;
    cnt += 1;
    for(int i=0;i<R-1;i++){
      tmp += cost[r[i]-1][r[i+1]-1];
    }
    ans = min(ans,tmp);
  }
  while( next_permutation(r.begin(),r.end()) );

  cout<<ans<<endl;
  // cout<<cnt<<endl;
  return 0;
}

File: 73/D.py
# encoding:utf-8
import copy
import numpy as np
import random
import bisect #bisect_left　これで二部探索の大小検索が行える
import math
import itertools

f_inf = float('inf')

N,M,R = map(int,input().split())
nodes = list(map(int,input().split()))
ways = []
for i in range(M):
    a,b,c = map(int,input().split())
    ways.append([a,b,c])

def warshall_floyd(n):
    dp = [[f_inf for i in range(N)] for j in range(N)]
    for i in range(n):
        dp[i][i] = 0
    for way in ways:
        dp[way[0]-1][way[1]-1] = way[2]
        dp[way[1]-1][way[0]-1] = way[2]

    for i in range(N):#頂点数　ノード数
        for j in range(N):#始点
            for k in range(N):#終点
                dp[j][k] = min(dp[j][k],dp[j][i]+dp[i][k])

    return dp


dp = warshall_floyd(N)
routes = list(itertools.permutations(nodes))

ans = f_inf
for route in routes:
#     route = list(print(route))
    tmp = 0
    for i in range(len(route)-1):
        # print(route[i])
        tmp += dp[route[i]-1][route[i+1]-1]
    ans = min(ans,tmp)


print(ans)
